syntax =
    "proto2";  // using proto2 to specify default-values
               // could do proto3 but then our code gets more responsibility
               // consider stronger support for json in proto3

package experimentserver;
import "google/protobuf/timestamp.proto";

enum Status {
  UNKNOWN = 0;
  STARTING = 1;
  RUNNING = 2;
  DONE = 3;
  FAILED = 4;
  CANCELLED = 5;
  PENDING = 6;
  WAITING = 7;
}

enum StepStatus {
  STEP_FAILURE = 0;
  STEP_PENDING = 1;
  STEP_RUNNING = 2;
  STEP_SUCCESS = 3;
  STEP_UNKNOWN = 4;
  STEP_WAITING = 5;
  STEP_TIMEOUT = 6;
}

enum EnsembleStatus {
  ENSEMBLE_STARTED = 0;
  ENSEMBLE_STOPPED = 1;
  ENSEMBLE_CANCELLED = 2;
  ENSEMBLE_FAILED = 3;
}

enum JobStatus {
  JOB_START = 0;
  JOB_RUNNING = 1;
  JOB_SUCCESS = 2;
  JOB_FAILURE = 3;
}

enum ExperimentStatus {
  EXPERIMENT_STARTED = 0;
  EXPERIMENT_SUCCEEDED = 1;
  EXPERIMENT_FAILED = 2;
  EXPERIMENT_CANCELLED = 3;
  EXPERIMENT_HOOK_STARTED = 4;
  EXPERIMENT_HOOK_ENDED = 5;
  EXPERIMENT_ANALYSIS_STARTED = 6;
  EXPERIMENT_ANALYSIS_ENDED = 7;
}

message DispatcherMessage {
  oneof object {
    Experiment experiment = 1;
    Ensemble ensemble = 2;
    Realization realization = 3;
    Step step = 4;
    Job job = 5;
  }
}

message ExperimentId {
  required string id = 1 [default = "experiment"];
}
message EnsembleId {
  required ExperimentId experiment = 1;
  required string id = 2
      [default =
           "ensemble"];  // an ensemble is identified by a string, not index
}
message RealizationId {
  required EnsembleId ensemble = 1;
  required uint64 realization = 2 [default = 0];
}
message StepId {
  required RealizationId realization = 1;
  required uint64 step = 2 [default = 0];  // replace with string?
}
message JobId {
  required StepId step = 1;
  required uint64 index = 2 [default = 0];
}

message Experiment {
  required ExperimentId id = 1;  // NOTE: in the message the id is optional
                                 // since it often will be streamed
  optional ExperimentStatus status = 2 [default = EXPERIMENT_STARTED];
  optional string message = 3;

  map<string, Ensemble> ensembles = 101;
}
message Ensemble {
  required EnsembleId id = 1;
  optional EnsembleStatus status = 2 [default = ENSEMBLE_STARTED];

  map<uint64, Realization> realizations = 101;
}
message Realization {
  required RealizationId id = 1;
  optional StepStatus status = 2 [default = STEP_UNKNOWN];
  optional bool active = 3
      [default = true];  // not derived - active/inactive by user or algorithm
  optional double start_time = 4;  // can be derived from step-list
  optional double end_time = 5;    // can be derived from step-list

  map<uint64, Step> steps = 101;
}
message Step {
  required StepId id = 1;
  optional StepStatus status = 2 [default = STEP_UNKNOWN];
  optional double start_time = 3;  // can be derived from job-list
  optional double end_time = 4;    // can be derived from job-list

  map<uint64, Job> jobs = 101;
}
message Job {
  required JobId id = 1;
  optional JobStatus status = 2 [default = JOB_START];
  optional google.protobuf.Timestamp start_time = 3;
  optional google.protobuf.Timestamp end_time = 4;
  optional string name = 6;
  optional string error = 7;
  optional string stdout = 8;
  optional string stderr = 9;

  optional uint64 current_memory = 10;
  optional uint64 max_memory = 11;
  optional int32 exit_code = 12;
}
